python/cudf/cudf/core/column/column.py

```diff
@@ -2318,285 +2333,69 @@ def as_column(
-                    dtype = cudf.dtype("str")
-                    pyarrow_array = pyarrow_array.cast(np_to_pa_dtype(dtype))
-
-                if (
-                    isinstance(arbitrary, pd.Index)
-                    and arbitrary.dtype == cudf.dtype("object")
-                    and (
-                        cudf.dtype(pyarrow_array.type.to_pandas_dtype())
-                        != cudf.dtype(arbitrary.dtype)
-                    )
-                ):
-                    raise MixedTypeError(
-                        "Cannot create column with mixed types"
-                    )
-
-                if (
-                    cudf.get_option("mode.pandas_compatible")
-                    and pa.types.is_integer(pyarrow_array.type)
-                    and pyarrow_array.null_count
-                ):
-                    pyarrow_array = pyarrow_array.cast("float64").fill_null(
-                        np.nan
+                    pd_dtype = dtype.to_pandas()
+                arbitrary = pd.Series(arbitrary, dtype=pd_dtype)
+            else:
+                if isinstance(dtype, np.dtype):
+                    typ = np_to_pa_dtype(dtype)
+                else:
+                    typ = dtype.to_arrow()
+                try:
+                    arbitrary = pa.array(
+                        arbitrary, type=typ, from_pandas=from_pandas
                     )
-
-                data = as_column(
-                    pyarrow_array,
-                    dtype=dtype,
-                    nan_as_null=nan_as_null,
-                )
-            except (pa.ArrowInvalid, pa.ArrowTypeError, TypeError) as e:
-                if isinstance(e, MixedTypeError):
-                    raise TypeError(str(e))
-                if is_categorical_dtype(dtype):
-                    sr = pd.Series(arbitrary, dtype="category")
-                    data = as_column(sr, nan_as_null=nan_as_null, dtype=dtype)
-                elif np_type == np.str_:
-                    sr = pd.Series(arbitrary, dtype="str")
-                    data = as_column(sr, nan_as_null=nan_as_null)
-                elif is_interval_dtype(dtype):
-                    sr = pd.Series(arbitrary, dtype="interval")
-                    data = as_column(sr, nan_as_null=nan_as_null, dtype=dtype)
-                elif (
-                    isinstance(arbitrary, Sequence)
-                    and len(arbitrary) > 0
-                    and any(
-                        cudf.utils.dtypes.is_column_like(arb)
-                        for arb in arbitrary
+                except (pa.ArrowInvalid, pa.ArrowTypeError):
+                    if not isinstance(dtype, np.dtype):
+                        dtype = dtype.to_pandas()
+                    arbitrary = pd.Series(arbitrary, dtype=dtype)
+            data = as_column(arbitrary, nan_as_null=nan_as_null)
+        else:
+            try:
+                arbitrary = pa.array(arbitrary, from_pandas=from_pandas)
```

Comment:
Do we want to go via arrow first and only then try and fall back to pandas? Or should we just go via pandas? I note this PR does not fix #14627 because I think we still go down the same (different) arrow vs pandas codepath.